pkg/join/join.go

```diff
@@ -0,0 +1,216 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package join
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/submariner-io/submariner-operator/internal/constants"
+	"github.com/submariner-io/submariner-operator/internal/image"
+	"github.com/submariner-io/submariner-operator/pkg/broker"
+	"github.com/submariner-io/submariner-operator/pkg/client"
+	"github.com/submariner-io/submariner-operator/pkg/deploy"
+	"github.com/submariner-io/submariner-operator/pkg/discovery/globalnet"
+	"github.com/submariner-io/submariner-operator/pkg/reporter"
+	"github.com/submariner-io/submariner-operator/pkg/secret"
+	"github.com/submariner-io/submariner-operator/pkg/version"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func SubmarinerCluster(brokerInfo *broker.Info, options *Options, clientProducer client.Producer,
```

Comment:
Wondering if we should rename this function. The _Submariner_ part seems redundant since we already know we're Submariner 😄 Some possibilities depending on perspective:

1. `join.Cluster`
2. `join.ToBroker`
3. `join.ToClusterSet`

I'm leaning towards 2. @Jaanki @skitt WDYT?

Reply:
I am leaning towards 1. It joins a cluster (to a broker or to a clusterSet) or lets go overboard with a 4th option `join.ClusterToBroker`.

Reply:
It depends on one's perspective and preference. `join.ClusterToBroker` would be the most complete. @skitt do you guys have any strong preference?

Reply:
Remembering Java, I like 1, because it names the result we’re after; the caller shouldn’t care that joining a cluster means joining a broker. However since Go doesn’t have polymorphism, and we’re explicitly passing in a `broker.Info`, `ClusterToBroker` makes sense to me — if we end up having other variants which don’t use a broker, we’ll say `ClusterToSomethingElse`.

Reply:
yeah makes sense. We'll go with `ClusterToBroker`. I think with just `Cluster`, it could be interpreted as "_joining to a cluster_" instead of "_joining a cluster to <something>_".